### PR TITLE
Feature/api create and delete relationship

### PR DESCRIPTION
**Pull request by mtakac2**

### CODE REVIEW

**Comment by myabc** [comment]
@hschink looks like the Karma tests are failing.

**Comment by hschink** [comment]
@myabc Sorry, not my branch (ATM). :smile: I talked to @mtakac about this.

File: lib/api/v3/relations/relations_api.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+        resources :relations do
+          params do
+            optional :to_id, desc: 'Id of related work package'
+            optional :relation_type, desc: 'Type of relationship'
+            optional :delay
+          end
+          post do
+            authorize(:manage_work_package_relations, context: @work_package.project)
+            declared_params = declared(params).reject{ |key, value| key.to_sym == :id || value.nil? }
```
Comment (hschink) [inline lib/api/v3/relations/relations_api.rb:14]: I don't understand why you remove `"id"` and null values here. Could you please enlighten me?

File: lib/api/v3/relations/relations_api.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+module API
+  module V3
+    module Relations
+      class RelationsAPI < Grape::API
+
+        resources :relations do
+          params do
+            optional :to_id, desc: 'Id of related work package'
+            optional :relation_type, desc: 'Type of relationship'
```
Comment (hschink) [inline lib/api/v3/relations/relations_api.rb:9]: I guess `:to_id` and `relation_type` are rather required?

File: lib/api/v3/relations/relations_api.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            optional :delay
+          end
+          post do
+            authorize(:manage_work_package_relations, context: @work_package.project)
+            declared_params = declared(params).reject{ |key, value| key.to_sym == :id || value.nil? }
+
+            relation = @work_package.new_relation.tap do |r|
+              r.to = WorkPackage.visible.find_by_id(declared_params[:to_id].match(/\d+/).to_s)
```
Comment (hschink) [inline lib/api/v3/relations/relations_api.rb:17]: Why do you need to match the id? Shouldn't `declared_params[:to_id]` be sufficient?

File: lib/api/v3/relations/relations_api.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+              r.relation_type = declared_params[:relation_type]
+              r.delay = declared_params[:delay_id]
+            end
+
+            if relation.valid?
+              model = ::API::V3::WorkPackages::RelationModel.new(relation)
+              representer = ::API::V3::WorkPackages::RelationRepresenter.new(model)
+              relation.save!
```
Comment (hschink) [inline lib/api/v3/relations/relations_api.rb]: I would rather write `if relation.valid? and relation.save`.

File: lib/services/create_relation.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+class Services::CreateRelation
```
Comment (hschink) [inline lib/services/create_relation.rb:1]: Is this actually used anywhere?

**Comment by hschink** [comment]
@mtakac I miss some tests for the new API endpoints and I guess you need to rebase the branch again.

File: lib/api/v3/relations/relations_api.rb (hunk 3 above)
Comment (myabc) [inline lib/api/v3/relations/relations_api.rb:17]: @mtakac why is this `match` here? Can't you do parameter validation in the `params` block above? e.g.

``` ruby
optional :to_id, desc: 'Id of related work package', type: String, regexp: /\d+/
# or better
optional :to_id, desc: 'Id of related work package', type: Integer

```

https://github.com/intridea/grape#parameter-validation-and-coercion

File: lib/services/create_relation.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      r.relation_type = attrs[:relation_type]
+      r.delay = attrs[:delay]
+    end
+  end
+
+  def run(success = -> {}, failure = -> {})
+    if @relation.save
+      success.(created: true)
+    else
+      binding.pry
```
Comment (myabc) [inline lib/services/create_relation.rb]: debugging statement shouldn't have been committed.

Comment (hschink) [inline lib/services/create_relation.rb]: @myabc :+1:

File: lib/api/v3/relations/relations_api.rb (hunk 2 above)
Comment (mtakac2) [inline lib/api/v3/relations/relations_api.rb:9]: Yes they are but I don't want to validate it here. I think the validation should be run on the object when we are trying to create it.

File: lib/api/v3/relations/relations_api.rb (hunk 3 above)
Comment (mtakac2) [inline lib/api/v3/relations/relations_api.rb:17]: @hschink I was just trying to use code as similar to the original as possible - no other reason.

Comment (mtakac2) [inline lib/api/v3/relations/relations_api.rb:17]: @myabc the problem is that I am not really sure how to customize response when validating type or format of a param. That's something we should probably take a look at, but I won't be able to do it this sprint.

**Comment by myabc** [comment]
@hschink thanks for the rebase/tidy up!

**Comment by hschink** [comment]
@myabc You're welcome!

File: lib/api/v3/relations/relations_api.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            else
+              raise ::API::Errors::Validation.new(relation)
+            end
+          end
+
+          namespace ':id' do
+            delete do
+              authorize(:manage_work_package_relations, context: @work_package.project)
+              relation = Relation.find(params[:id])
+              relation.delete
```
Comment (myabc) [inline lib/api/v3/relations/relations_api.rb]: `#destroy` or `#delete`? Do we really want to skip callbacks (and ignore the `:dependent` option) here? http://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-delete

Why not `Relation.destroy(params[:id])`?

File: lib/api/v3/relations/relations_api.rb (hunk 3 above)
Comment (myabc) [inline lib/api/v3/relations/relations_api.rb:17]: @mtakac you mean customising the error response?

File: lib/api/v3/relations/relations_api.rb (hunk 2 above)
Comment (myabc) [inline lib/api/v3/relations/relations_api.rb:9]: I'd return `400: Bad Request` if these params are missing and if more specific validations fail, then `422`.

Comment (mtakac2) [inline lib/api/v3/relations/relations_api.rb:9]: @myabc I actually don't think we should return `400`. If the params are missing, object is invalid, therefore `422` with errors messages in the body.

File: lib/api/v3/relations/relations_api.rb (hunk 3 above)
Comment (mtakac2) [inline lib/api/v3/relations/relations_api.rb:17]: @myabc Yes exactly. I am just looking into that. Got any tips?

**Comment by hschink** [comment]
@mtakac, @myabc I rebased the branch.

**Comment by myabc** [comment]
Rebased. Should now be mergeable.

**Pull request merged**